File: mm-php/src/io/metamessage/jsonc/Jsonc.php
<?php

namespace io\metamessage\jsonc;

use io\metamessage\ir\Node;
use io\metamessage\ir\NodeObject;
use io\metamessage\ir\NodeArray;
use io\metamessage\ir\NodeScalar;
use io\metamessage\ir\Field;

class_alias(NodeObject::class, 'io\\metamessage\\jsonc\\JsoncObject');
class_alias(NodeArray::class, 'io\\metamessage\\jsonc\\JsoncArray');
class_alias(NodeScalar::class, 'io\\metamessage\\jsonc\\JsoncValue');
class_alias(Field::class, 'io\\metamessage\\jsonc\\JsoncField');

function parseJsonc(string $s): Node
{
    return Jsonc::ParseFromString($s);
}

class Jsonc
{
    public static function ToJSONC(?Node $n): string
    {
        if ($n === null) {
            return '';
        }

        $printer = new JsoncPrinter();
        return $printer::ToJSONC($n);
    }

    public static function ParseFromString(string $s): Node
    {
        $scanner = new JsoncScanner($s);
        $toks = [];
        while (true) {
            $t = $scanner->nextToken();
            $toks[] = $t;
            if ($t->type === JsoncTokenType::EOF) {
                break;
            }
        }

        $parser = new JsoncParser($toks);
        return $parser->parse();
    }

    public static function Print(Node $n): void
    {
        echo self::ToJSONC($n) . "\n";
    }
}

File: mm-kt/src/main/kotlin/io/github/metamessage/jsonc/Jsonc.kt
package io.github.metamessage.jsonc

import io.github.metamessage.ir.Node

class JsoncException(message: String) : Exception(message)

fun parseFromJsonc(jsonc: String): Node {
    val scanner = JsoncScanner(jsonc)
    val tokens = mutableListOf<JsoncToken>()
    while (true) {
        val t = scanner.nextToken()
        tokens.add(t)
        if (t.type == JsoncTokenType.EOF) break
    }
    val parser = JsoncParser(tokens)
    return parser.parse() ?: throw JsoncException("failed to parse JSONC")
}

fun printJsonc(node: Node?): String {
    if (node == null) return ""
    val sb = StringBuilder()
    writeLeadingComments(sb, node.tag, 0)
    sb.append(JsoncPrinter.toString(node))
    return sb.toString()
}

private fun writeLeadingComments(
        sb: StringBuilder,
        tag: io.github.metamessage.ir.Tag?,
        indentLevel: Int
) {
    val tagStr = tag?.toString() ?: ""
    if (tagStr.isNotEmpty()) {
        sb.append("// mm: $tagStr\n")
    }
}

fun toJsonc(node: Node?): String {
    return printJsonc(node)
}
